beacon-chain/db/kv/state.go

```diff
@@ -99,30 +114,112 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
 		return errors.New("nil state")
 	}
 	multipleEncs := make([][]byte, len(states))
+	validatorsEntries := make(map[string][]byte)              // It's a map to make sure that you store only new validator entries.
+	validatorKeys := make([][]byte, len(states))              // For every state, this stores a compressed list of validator keys.
+	realValidators := make([][]*ethpb.Validator, len(states)) // It's temporary structure to restore state in memory after persisting it.
 	for i, st := range states {
 		pbState, err := v1.ProtobufBeaconState(st.InnerStateUnsafe())
 		if err != nil {
 			return err
 		}
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
```

Comment:
This came up when looking through this function, I realised that it might be difficult inserting all these conditionals here to account for historical state representation this way. How about having a completely new Save State method for historical representation ?

```go 

func(s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconState, blockRoots [][32]byte) error {
		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
                  return s.SaveStatesEfficient(ctx context.Context, states []iface.ReadOnlyBeaconState, blockRoots [][32]byte)
}
.
.
.
}
```
That way you can keep your method 'pure' , you won't have to constantly mess around with flags here multiple times.